README.md
# Arquebuse client

 [![Linux-Build](https://github.com/arquebuse/arquebuse-ps/workflows/Linux-Build/badge.svg)](https://github.com/arquebuse/arquebuse-ps/actions?query=workflow%3ALinux-Build) [![MacOS-Build](https://github.com/arquebuse/arquebuse-ps/workflows/MacOS-Build/badge.svg)](https://github.com/arquebuse/arquebuse-ps/actions?query=workflow%3AMacOS-Build) [![Windows-Build](https://github.com/arquebuse/arquebuse-ps/workflows/Windows-Build/badge.svg)](https://github.com/arquebuse/arquebuse-ps/actions?query=workflow%3AWindows-Build) [![Windows-PS51-Build](https://github.com/arquebuse/arquebuse-ps/workflows/Windows-PS51-Build/badge.svg)](https://github.com/arquebuse/arquebuse-ps/actions?query=workflow%3AWindows-PS51-Build)

Arquebuse PowerShell Client

The client is distributed as a PowerShell module and works for:

* Windows PowerShell 5.1
* Powershell 7 (tested on Windows, Linux and MacOS)

For more information about [Arquebuse](https://arquebuse.io), an email infrastructure testing tool, please visit [Arquebuse website](https://arquebuse.io) or the [main project page](https://github.com/arquebuse/arquebuse).

## How to build

To test and build the module

    cd src
    Invoke-Build

## Thanks to

* [Jake Morrison](https://techthoughts.info) for the [Catesta PowerShell module project generator](https://github.com/techthoughts2/Catesta)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.2]

### Added

- Initial (pre)release.


src/Tests/Unit/Arquebuse-Module.Tests.ps1
#-------------------------------------------------------------------------
Set-Location -Path $PSScriptRoot
#-------------------------------------------------------------------------
$ModuleName = 'Arquebuse'
$PathToManifest = [System.IO.Path]::Combine('..', '..', $ModuleName, "$ModuleName.psd1")
$PathToModule = [System.IO.Path]::Combine('..', '..', $ModuleName, "$ModuleName.psm1")
#-------------------------------------------------------------------------
Describe 'Module Tests' -Tag Unit {
    Context "Module Tests" {
        It 'Passes Test-ModuleManifest' {
            Test-ModuleManifest -Path $PathToManifest | Should Not BeNullOrEmpty
            $? | Should Be $true
        }
        It 'root module Arquebuse.psm1 should exist' {
            $PathToModule | Should Exist
            $? | Should Be $true
        }
        It 'manifest should contain Arquebuse.psm1' {
            $PathToManifest |
                Should -FileContentMatchExactly "Arquebuse.psm1"
        }
    }
}


src/Tests/Unit/Private-Function.Tests.ps1
#-------------------------------------------------------------------------
Set-Location -Path $PSScriptRoot
#-------------------------------------------------------------------------
$PathToPrivateFunction = [System.IO.Path]::Combine('..', '..', $ModuleName, 'Private', 'Private.ps1')
$PathToPsd1ConfigFile  = [System.IO.Path]::Combine('..', 'Resources', 'config.psd1')
$PathToFakePsd1File  = [System.IO.Path]::Combine('..', 'Resources', 'fake.psd1')
$PathToFakeJsonFile  = [System.IO.Path]::Combine('..', 'Resources', 'fake.json')
$PathToYamlConfigFile  = [System.IO.Path]::Combine('..', 'Resources', 'config.yaml')
$PathToHome  = [System.IO.Path]::Combine('..', 'Resources', 'Home')
#-------------------------------------------------------------------------
. $PathToPrivateFunction
#-------------------------------------------------------------------------
$WarningPreference = "SilentlyContinue"
#-------------------------------------------------------------------------

function Assert-HashtableEquality($test, $expected) {
    $test.Keys | Should -HaveCount $expected.Keys.Count
    $test.Keys | ForEach-Object {$test[$_] | Should -Be $expected[$_]}
}

Describe 'Arquebuse Private Function Tests' -Tag Unit {
    Context 'GetHome' {
        It 'Returns the HOME folder' {
            GetHome | Should -Be $HOME
        }
    }

    Context 'GetCommonApiParameters' {
        Mock GetHome {return 'nowhere'}

        # Avoid warnings emitted by bad config file to be detected by Pester
        Mock Write-Warning {}

        It 'Returns the default values' {
            $test = GetCommonApiParameters
            $test.BaseUrl | Should -Be 'https://localhost'
            $test.SkipCertificateCheck | Should -Be $false
            $test.ApiKey | Should -Be ''
        }

        It 'Returns the default values when a bad JSON config file is provided' {
            $test = GetCommonApiParameters -ConfigPath $PathToFakeJsonFile
            $test.BaseUrl | Should -Be 'https://localhost'
            $test.SkipCertificateCheck | Should -Be $false
            $test.ApiKey | Should -Be ''
        }

        It 'Returns the default values when a bad PSD1 config file is provided' {
            $test = GetCommonApiParameters -ConfigPath $PathToFakePsd1File
            $test.BaseUrl | Should -Be 'https://localhost'
            $test.SkipCertificateCheck | Should -Be $false
            $test.ApiKey | Should -Be ''
        }

        It 'Returns the default values when an unsupported config file is provided' {
            $test = GetCommonApiParameters -ConfigPath $PathToYamlConfigFile
            $test.BaseUrl | Should -Be 'https://localhost'
            $test.SkipCertificateCheck | Should -Be $false
            $test.ApiKey | Should -Be ''
        }

        It 'Returns values from a specified PSD1 config file' {
            $test = GetCommonApiParameters -ConfigPath $PathToPsd1ConfigFile
            $test.BaseUrl | Should -Be 'fake-url'
            $test.SkipCertificateCheck | Should -Be $true
            $test.ApiKey | Should -Be 'fake-api-key'
        }

        It 'Returns values from a specified PSD1 config file + overrided ones' {
            $test = GetCommonApiParameters -ConfigPath $PathToPsd1ConfigFile -BaseUrl 'my-url'
            $test.BaseUrl | Should -Be 'my-url'
            $test.SkipCertificateCheck | Should -Be $true
            $test.ApiKey | Should -Be 'fake-api-key'
        }

        Mock GetHome {return $PathToHome}

        It 'Returns values from default json config file' {
            $test = GetCommonApiParameters
            $test.BaseUrl | Should -Be 'another-fake-url'
            $test.SkipCertificateCheck | Should -Be $false
            $test.ApiKey | Should -Be 'another-fake-api-key'
        }

        It 'Returns values from default json config file' {
            $test = GetCommonApiParameters -APIKey 'my-api-key' -SkipCertificateCheck $true
            $test.BaseUrl | Should -Be 'another-fake-url'
            $test.SkipCertificateCheck | Should -Be $true
            $test.ApiKey | Should -Be 'my-api-key'
        }

    }

    Context 'InvokeApi' {
        Mock Invoke-RestMethod {return 'test'}

        $ApiParameters = @{
            BaseUrl = 'fake-url'
            ApiKey  = 'fake-api-key'
            Object  = 'test'
        }

        It 'Returns the requested object' {
            $result = InvokeApi @ApiParameters
            $result.Success | Should -Be $true
            $result.Message | Should -Be ''
            $result.Data    | Should -Be 'test'
        }

        # Mock errors only with PowerShell 6 and beyond
        if ($PSVersionTable.PSVersion.Major -ge 6) {
            Mock Invoke-RestMethod {
                $httpResponseMessage = [System.Net.Http.HttpResponseMessage]::new(404)
                throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("404 page not found", $httpResponseMessage)
            }

            It 'Returns null when API returns a 404' {
                $result = InvokeApi @ApiParameters
                $result.Success | Should -Be $true
                $result.Message | Should -Be ''
                $result.Data    | Should -Be $null
            }

            Mock Invoke-RestMethod {
                $innerException = [System.SystemException]::new("The remote certificate is invalid according to the validation procedure.")
                throw [System.Net.Http.HttpRequestException]::new("The SSL connection could not be established, see inner exception.", $innerException)
            }

            It 'Reports bad SSL certificates' {
                $result = InvokeApi @ApiParameters
                $result.Success | Should -Be $false
                $result.Message | Should -BeLike '*-SkipCertificateCheck*'
                $result.Data    | Should -Be $null
            }

            Mock Invoke-RestMethod {
                $httpResponseMessage = [System.Net.Http.HttpResponseMessage]::new(401)
                throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Not Authorized", $httpResponseMessage)
            }

            It 'Advices to check API-Key' {
                $result = InvokeApi @ApiParameters
                $result.Success | Should -Be $false
                $result.Message | Should -BeLike '*API-Key*'
                $result.Data    | Should -Be $null
            }

            Mock Invoke-RestMethod {
                $httpResponseMessage = [System.Net.Http.HttpResponseMessage]::new(500)
                throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Internal error", $httpResponseMessage)
            }

            It 'Reports unknown status codes' {
                $result = InvokeApi @ApiParameters
                $result.Success | Should -Be $false
                $result.Message | Should -BeLike '*Status Code: 500*'
                $result.Data    | Should -Be $null
            }

            Mock Invoke-RestMethod {
                throw "Find Me"
            }

            It 'Reports unknown exceptions' {
                $result = InvokeApi @ApiParameters
                $result.Success | Should -Be $false
                $result.Message | Should -BeLike '*Find Me*'
                $result.Data    | Should -Be $null
            }

            Mock Invoke-RestMethod {
                throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Find Me", $null)
            }

            It 'Reports unknown HTTP exceptions' {
                $result = InvokeApi @ApiParameters
                $result.Success | Should -Be $false
                $result.Message | Should -BeLike '*Find Me*'
                $result.Data    | Should -Be $null
            }

            Mock Invoke-RestMethod {
                $innerException = [System.SystemException]::new("Find Me")
                throw [System.Net.Http.HttpRequestException]::new("Unknown error", $innerException)
            }

            It 'Reports unknown HTTP inner exceptions' {
                $result = InvokeApi @ApiParameters
                $result.Success | Should -Be $false
                $result.Message | Should -BeLike '*Find Me*'
                $result.Data    | Should -Be $null
            }
        }
    }

    Context 'GetApiVersion' {
        Mock Invoke-RestMethod {return 'version'}

        It 'Returns the remote API version' {
            $commonApiParameters = @{
                BaseUrl              = 'fake-url'
                ApiKey               = 'fake-api-key'
            }
            GetApiVersion -CommonApiParameters $commonApiParameters | Should -Be 'version'
        }

        Mock Invoke-RestMethod {
            throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Unknow error", $null)
        }

        It 'Throws when API is not available' {
            $commonApiParameters = @{
                BaseUrl              = 'fake-url'
                ApiKey               = 'fake-api-key'
            }
            { GetApiVersion -CommonApiParameters $commonApiParameters } | Should -Throw
        }
    }

    Context 'GetInbound' {
        Mock Invoke-RestMethod {return [PSCustomObject]@{
            id        = "1fuzRupHO5euYJCaKOFovsfY3Ck"
            timestamp = "2020-08-10T18:06:00.911185495Z"
            client    = "127.0.0.1:33472"
            from      = "someone@arquebuse.org"
            to        = "someone.else@arquebuse.org"
            subject   = "Hello from Arquebuse UI !!!"
        }}

        It 'Returns the list of inbound emails' {
            $commonApiParameters = @{
                BaseUrl              = 'fake-url'
                ApiKey               = 'fake-api-key'
            }
            $result = GetInbound -CommonApiParameters $commonApiParameters
            $result.id | Should -Be "1fuzRupHO5euYJCaKOFovsfY3Ck"
            $result.timestamp | Should -Be "2020-08-10T18:06:00.911185495Z"
            $result.client | Should -Be "127.0.0.1:33472"
            $result.from | Should -Be "someone@arquebuse.org"
            $result.to | Should -Be "someone.else@arquebuse.org"
            $result.subject | Should -Be "Hello from Arquebuse UI !!!"
        }

        Mock Invoke-RestMethod {
            throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Unknow error", $null)
        }

        It 'Throws when API is not available' {
            $commonApiParameters = @{
                BaseUrl              = 'fake-url'
                ApiKey               = 'fake-api-key'
            }
            { GetInbound -CommonApiParameters $commonApiParameters } | Should -Throw
        }
    }

    Context 'GetOutbound' {
        Mock Invoke-RestMethod {return [PSCustomObject]@{
            id        = "1fuzRupHO5euYJCaKOFovsfY3Ck"
            timestamp = "2020-08-10T18:06:00.911185495Z"
            client    = "127.0.0.1:33472"
            from      = "someone@arquebuse.org"
            to        = "someone.else@arquebuse.org"
            subject   = "Hello from Arquebuse UI !!!"
        }}

        It 'Returns the list of outbound emails' {
            $commonApiParameters = @{
                BaseUrl              = 'fake-url'
                ApiKey               = 'fake-api-key'
            }
            $result = GetOutbound -CommonApiParameters $commonApiParameters
            $result.id | Should -Be "1fuzRupHO5euYJCaKOFovsfY3Ck"
            $result.timestamp | Should -Be "2020-08-10T18:06:00.911185495Z"
            $result.client | Should -Be "127.0.0.1:33472"
            $result.from | Should -Be "someone@arquebuse.org"
            $result.to | Should -Be "someone.else@arquebuse.org"
            $result.subject | Should -Be "Hello from Arquebuse UI !!!"
        }

        Mock Invoke-RestMethod {
            throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Unknow error", $null)
        }

        It 'Throws when API is not available' {
            $commonApiParameters = @{
                BaseUrl              = 'fake-url'
                ApiKey               = 'fake-api-key'
            }
            { GetOutbound -CommonApiParameters $commonApiParameters } | Should -Throw
        }
    }

    Context 'ToEmailObject' {
        It 'Converts a valid email result into an ArquebuseEmail object' {
            $object = [PSCustomObject]@{
                id        = "1fuzRupHO5euYJCaKOFovsfY3Ck"
                timestamp = "2020-08-10T18:06:00.911185495Z"
                client    = "127.0.0.1:33472"
                from      = "someone@arquebuse.org"
                to        = "someone.else@arquebuse.org"
                subject   = "Hello from Arquebuse UI !!!"
                status    = "Sent"
            }
            $result = ToEmailObject -InputObject $object
            $result.Date.ToUniversalTime().ToString("yyMMdd HH:mm:ss") | Should -Be '200810 18:06:00'
            $result.ID          | Should -Be "1fuzRupHO5euYJCaKOFovsfY3Ck"
            $result.Client      | Should -Be "127.0.0.1"
            $result.ClientPort  | Should -Be 33472
            $result.From        | Should -Be "someone@arquebuse.org"
            $result.To          | Should -Be "someone.else@arquebuse.org"
            $result.Subject     | Should -Be "Hello from Arquebuse UI !!!"
            $result.Status      | Should -Be "Sent"
        }

        It 'Returns default values for absent object values' {
            $object = [PSCustomObject]@{}
            $result = ToEmailObject -InputObject $object
            $result.Date.Year   | Should -Be $(Get-Date).Year
            $result.Date.Month  | Should -Be $(Get-Date).Month
            $result.Date.Day    | Should -Be $(Get-Date).Day
            $result.Date.Hour   | Should -Be $(Get-Date).Hour
            $result.ID          | Should -Be ""
            $result.Client      | Should -Be "none"
            $result.ClientPort  | Should -Be 0
            $result.From        | Should -Be ""
            $result.To          | Should -Be ""
            $result.Subject     | Should -Be ""
            $result.Status      | Should -Be "RECEIVED"
        }

        It 'Returns default values for incorrect object values' {
            $object = [PSCustomObject]@{
                client    = "error"
            }
            $result = ToEmailObject -InputObject $object
            $result.Client      | Should -Be "error"
            $result.ClientPort  | Should -Be 0
        }
    }
}

src/Tests/Unit/Arquebuse-Function.Tests.ps1
#-------------------------------------------------------------------------
Set-Location -Path $PSScriptRoot
#-------------------------------------------------------------------------
$ModuleName = 'Arquebuse'
$PathToManifest = [System.IO.Path]::Combine('..', '..', $ModuleName, "$ModuleName.psd1")
#-------------------------------------------------------------------------
if (Get-Module -Name $ModuleName -ErrorAction 'SilentlyContinue') {
    #if the module is already in memory, remove it
    Remove-Module -Name $ModuleName -Force
}
Import-Module $PathToManifest -Force
#-------------------------------------------------------------------------
$WarningPreference = "SilentlyContinue"
#-------------------------------------------------------------------------
#Import-Module $moduleNamePath -Force

InModuleScope 'Arquebuse' {
    #-------------------------------------------------------------------------
    $WarningPreference = "SilentlyContinue"
    #-------------------------------------------------------------------------

    Describe 'Arquebuse Public Function Tests' -Tag Unit {
        Context 'Get-ArquebuseVersion' {
            Mock Invoke-RestMethod {return 'version'}

            It 'Returns the remote API version' {
                $commandParameters = @{
                    BaseUrl              = 'fake-url'
                    ApiKey               = 'fake-api-key'
                    SkipCertificateCheck = $false
                }
                Get-ArquebuseVersion @commandParameters | Should -Be 'version'
            }

            Mock Invoke-RestMethod {
                throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Unknow error", $null)
            }

            It 'Throws when API is not available' {
                $commandParameters = @{
                    BaseUrl              = 'fake-url'
                    ApiKey               = 'fake-api-key'
                    SkipCertificateCheck = $false
                }
                { GetApiVersion @commandParameters } | Should -Throw
            }
        }

        Context 'Get-ArquebuseInbound' {
            Mock Invoke-RestMethod {return [PSCustomObject]@{
                id        = "1fuzRupHO5euYJCaKOFovsfY3Ck"
                timestamp = "2020-08-10T18:06:00.911185495Z"
                client    = "127.0.0.1:33472"
                from      = "someone@arquebuse.org"
                to        = "someone.else@arquebuse.org"
                subject   = "Hello from Arquebuse UI !!!"
            }}

            It 'Returns the list of inbound emails' {
                $commandParameters = @{
                    BaseUrl              = 'fake-url'
                    ApiKey               = 'fake-api-key'
                    SkipCertificateCheck = $false
                }

                $result = Get-ArquebuseInbound @commandParameters
                $result.Date.ToUniversalTime().ToString("yyMMdd HH:mm:ss") | Should -Be '200810 18:06:00'
                $result.ID          | Should -Be "1fuzRupHO5euYJCaKOFovsfY3Ck"
                $result.Client      | Should -Be "127.0.0.1"
                $result.ClientPort  | Should -Be 33472
                $result.From        | Should -Be "someone@arquebuse.org"
                $result.To          | Should -Be "someone.else@arquebuse.org"
                $result.Subject     | Should -Be "Hello from Arquebuse UI !!!"
                $result.Status      | Should -Be "RECEIVED"
            }

            Mock Invoke-RestMethod {
                throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Unknow error", $null)
            }

            It 'Throws when API is not available' {
                $commandParameters = @{
                    BaseUrl              = 'fake-url'
                    ApiKey               = 'fake-api-key'
                    SkipCertificateCheck = $false
                }

                { Get-ArquebuseInbound @commandParameters } | Should -Throw
            }
        }

        Context 'Get-ArquebuseOutbound' {
            Mock Invoke-RestMethod {return [PSCustomObject]@{
                id        = "1fuzRupHO5euYJCaKOFovsfY3Ck"
                timestamp = "2020-08-10T18:06:00.911185495Z"
                client    = "127.0.0.1:33472"
                from      = "someone@arquebuse.org"
                to        = "someone.else@arquebuse.org"
                subject   = "Hello from Arquebuse UI !!!"
                status    = "Failed"
            }}

            It 'Returns the list of outbound emails' {
                $commandParameters = @{
                    BaseUrl              = 'fake-url'
                    ApiKey               = 'fake-api-key'
                    SkipCertificateCheck = $false
                }

                $result = Get-ArquebuseOutbound @commandParameters
                $result.Date.ToUniversalTime().ToString("yyMMdd HH:mm:ss") | Should -Be '200810 18:06:00'
                $result.ID          | Should -Be "1fuzRupHO5euYJCaKOFovsfY3Ck"
                $result.Client      | Should -Be "127.0.0.1"
                $result.ClientPort  | Should -Be 33472
                $result.From        | Should -Be "someone@arquebuse.org"
                $result.To          | Should -Be "someone.else@arquebuse.org"
                $result.Subject     | Should -Be "Hello from Arquebuse UI !!!"
                $result.Status      | Should -Be "Failed"
            }

            Mock Invoke-RestMethod {
                throw [Microsoft.PowerShell.Commands.HttpResponseException]::new("Unknow error", $null)
            }

            It 'Throws when API is not available' {
                $commandParameters = @{
                    BaseUrl              = 'fake-url'
                    ApiKey               = 'fake-api-key'
                    SkipCertificateCheck = $false
                }

                { Get-ArquebuseOutbound @commandParameters } | Should -Throw
            }
        }
    }
}


src/Bootstrap.ps1
# Bootstrap dependencies

# Avoid progress bar artefacts on reports
$ProgressPreference = 'SilentlyContinue'

# Fail on error
$ErrorActionPreference = 'Stop'

# https://docs.microsoft.com/powershell/module/packagemanagement/get-packageprovider
Get-PackageProvider -Name Nuget -ForceBootstrap | Out-Null

# https://docs.microsoft.com/powershell/module/powershellget/set-psrepository
Set-PSRepository -Name PSGallery -InstallationPolicy Trusted

#Include: Settings
$moduleName = 'Arquebuse'
$moduleSettingsPath = Join-Path -Path $PSScriptRoot -ChildPath "$moduleName.Settings.ps1"
. $moduleSettingsPath

'Installing PowerShell Modules'
foreach ($module in $requiredModules) {
    $installSplat = @{
        Name            = $module.ModuleName
        RequiredVersion = $module.ModuleVersion
        Repository      = 'PSGallery'
        Force           = $true
        ErrorAction     = 'Stop'
    }
    try {
        Install-Module @installSplat
        Import-Module -Name $module.ModuleName -RequiredVersion $module.ModuleVersion -Force

        '  - Successfully installed {0} version {1}' -f $module.ModuleName, $(get-module -Name $module.ModuleName).Version.ToString()
    }
    catch {
        $message = 'Failed to install {0}' -f $module.ModuleName
        "  - $message"
        throw $message
    }
}


src/Arquebuse.Settings.ps1

# List of PowerShell Modules required for the build
$requiredModules = [System.Collections.ArrayList]::new()

# https://github.com/pester/Pester
$null = $requiredModules.Add(([PSCustomObject]@{
            ModuleName    = 'Pester'
            ModuleVersion = '4.9.0'
        }))
# https://github.com/nightroman/Invoke-Build
$null = $requiredModules.Add(([PSCustomObject]@{
            ModuleName    = 'InvokeBuild'
            ModuleVersion = '5.6.0'
        }))
# https://github.com/PowerShell/PSScriptAnalyzer
$null = $requiredModules.Add(([PSCustomObject]@{
            ModuleName    = 'PSScriptAnalyzer'
            ModuleVersion = '1.19.0'
        }))
# https://github.com/PowerShell/platyPS
# older version used due to: https://github.com/PowerShell/platyPS/issues/457
$null = $requiredModules.Add(([PSCustomObject]@{
            ModuleName    = 'platyPS'
            ModuleVersion = '0.12.0'
        }))

src/Arquebuse/Private/Private.ps1
# Function to invoke Arquebuse API and process returned object or errors
function InvokeApi {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [string]$BaseUrl,

        [Parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [string]$ApiKey,

        [Parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [string]$Object,

        [Parameter()]
        [string]$SubObject = '',

        [Parameter()]
        [string]$ID = '',

        [Parameter()]
        [string]$Version = 'v1',

        [Parameter()]
        [ValidateSet('GET', 'POST', 'PUT', 'PATCH', 'DELETE')]
        [string]$Method = 'GET',

        [Parameter()]
        [bool]$SkipCertificateCheck = $false
    )

    # As SkipCertificateCheck is not available in PowerShell version before 6
    if ($PSVersionTable.PSVersion.Major -lt 6) {
        # Save Certificate Policy
        $certificatePolicy = [System.Net.ServicePointManager]::CertificatePolicy

        if ($SkipCertificateCheck) {
            if (-not("TrustAllCertsPolicy" -as [type])) {
                Add-Type @"
using System.Net;
using System.Security.Cryptography.X509Certificates;
public class TrustAllCertsPolicy : ICertificatePolicy {
    public bool CheckValidationResult(
        ServicePoint srvPoint, X509Certificate certificate,
        WebRequest request, int certificateProblem) {
        return true;
    }
}
"@
            }

            # Replace Certificate Policy
            [System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy
        }

        $SkipCertificateCheckParameter = @{}
    } else {
        $SkipCertificateCheckParameter = @{
            SkipCertificateCheck = $SkipCertificateCheck
        }
    }

    $uri = ($BaseUrl, 'api', $Version, $Object, $SubObject, $ID | Where-Object { $_ }) -join '/'

    $data    = $null
    $message = ''
    $success = $true
    try {
        $data = Invoke-RestMethod -Uri $uri -Method $Method -Headers @{'X-API-Key' = $ApiKey} @SkipCertificateCheckParameter
    } catch {
        $success = $false
        $exception = $_.Exception
        $message = "Error while querying '$uri'. Message: "
        if ($exception.message -like '*SSL*') {
            Write-Verbose $exception.Message
            if ($exception.InnerException) { Write-Verbose $exception.InnerException.Message }
            $message += "Bad SSL certificate - Please verify your SSL configuration or use -SkipCertificateCheck parameter"
        } else {
            if ($exception.Response) {
                $statusCode = [Int32]$exception.Response.StatusCode
                switch ($statusCode) {
                    401 {
                        $message += "Unauthorized - Please verify your API-Key"
                    }
                    404 {
                        $message = ''
                        $success = $true
                    }
                    default {
                        $message += "Status Code: $statuscode - $($exception.Message)"
                    }
                }
            } else {
                if ($exception.InnerException) {
                    $message += $exception.InnerException.Message
                } else {
                    $message += $exception.Message
                }
            }
        }
    } finally {
        # Restore previously saved Certificate Policy
        if ($PSVersionTable.PSVersion.Major -lt 6) {
            [System.Net.ServicePointManager]::CertificatePolicy = $certificatePolicy
        }
    }

    return [PSCustomObject]@{
        Success = $success
        Data    = $data
        Message = $message
    }
}

# Function to get API version object from remote API
function GetApiVersion {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory = $true)]
        [object]$CommonApiParameters
    )

    $result = (InvokeApi -Object 'system' -SubObject 'info' @CommonApiParameters)

    if ($result.Success) {
        return $result.Data
    } else {
        Throw $result.message
    }
}

# Function to list inbound emails
function GetInbound {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory = $true)]
        [object]$CommonApiParameters
    )

    $result = (InvokeApi -Object 'inbound' @CommonApiParameters)

    if ($result.Success) {
        return $result.Data
    } else {
        Throw $result.message
    }
}

# Function to list outbound emails
function GetOutbound {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory = $true)]
        [object]$CommonApiParameters
    )

    $result = (InvokeApi -Object 'outbound' @CommonApiParameters)

    if ($result.Success) {
        return $result.Data
    } else {
        Throw $result.message
    }
}

# Function to get HOME path (usefull in unit tests)
function GetHome { return $HOME }

# Function to get common API parameters from command line arguments or from configuration
function GetCommonApiParameters {
    [CmdletBinding()]
    param (
        [Parameter()]
        [string]$ConfigPath,

        [Parameter()]
        [string]$BaseUrl,

        [Parameter()]
        [string]$ApiKey,

        [Parameter()]
        [bool]$SkipCertificateCheck = $false
    )

    $parameters = @{
        BaseUrl              = 'https://localhost'
        SkipCertificateCheck = $false
        ApiKey               = ''
    }

    if ($ConfigPath) {
        $configPathCandidate = @( $ConfigPath )
    } else {
        $configPathCandidate = @(
            $(Join-Path -Path $(GetHome) -ChildPath '.arquebuse.psd1')
            $(Join-Path -Path $(GetHome) -ChildPath '.arquebuse.json')
        )
    }

    $config = $null
    foreach ($path in $configPathCandidate) {
        if (Test-Path -Path $path -PathType Leaf -ErrorAction SilentlyContinue) {
            $extension = $path -replace '.*\.', ''
            switch ($extension) {
                'psd1' {
                    try {
                        $config = Import-PowerShellDataFile -Path $path -ErrorAction 'SilentlyContinue' -ErrorVariable ImportError | ConvertTo-Json -Depth 99 | ConvertFrom-Json
                        if ($ImportError) {
                            $config = $null
                            Throw $ImportError
                        } else {
                            Write-Verbose "Successfully loaded config file '$path'"
                        }
                    } catch {
                        Write-Warning "Failed to load config file '$path'. Message: $($_.Exception.Message)"
                    }
                }
                'json' {
                    try {
                        $config = Get-Content -Path $path -Raw | ConvertFrom-Json
                        Write-Verbose "Successfully loaded config file '$path'"
                    } catch {
                        Write-Warning "Failed to load config file '$path'. Message: $($_.Exception.Message)"
                    }
                }
                default {
                    Write-Warning "Cannot load config file '$path'. Supported extensions are JSON and PSD1"
                }
            }
        }

        if ($config) {
            break
        }
    }

    if ($config -and (Get-Member -InputObject $config -Name 'DefaultServer')) {
        $defaultServer = $config.DefaultServer
        if (Get-Member -InputObject $defaultServer -Name 'BaseUrl') {
            $parameters.BaseUrl = $defaultServer.BaseUrl
        }

        if (Get-Member -InputObject $defaultServer -Name 'ApiKey') {
            $parameters.ApiKey = $defaultServer.ApiKey
        }

        if (Get-Member -InputObject $defaultServer -Name 'SkipCertificateCheck') {
            $parameters.SkipCertificateCheck = $defaultServer.SkipCertificateCheck
        }
    }

    if ($BaseUrl) {
        $parameters.BaseUrl = $BaseUrl
    }

    if ($ApiKey) {
        $parameters.ApiKey = $ApiKey
    }

    if ($SkipCertificateCheck) {
        $parameters.SkipCertificateCheck = $SkipCertificateCheck
    }

    return $parameters

}

# Function to convert a PSCustomObject into an ArquebuseEmail object
function ToEmailObject {
    [cmdletbinding()]
    param (
        [Parameter(ValueFromPipeline)]
        [object]$InputObject
    )

    Process {
        if (Get-Member -InputObject $InputObject -Name 'client') {
            $client = $InputObject.client
            if ($client -match "^(?<client>[^:]+):(?<port>[0-9]+)$") {
                $client = $Matches['client']
                $clientPort = [int]$Matches['port']
            } else {
                $clientPort = 0
            }
        } else {
            $client     = "none"
            $clientPort = 0
        }

        if (Get-Member -InputObject $InputObject -Name 'from') {
            $from = $InputObject.from
        } else {
            $from = ""
        }

        if (Get-Member -InputObject $InputObject -Name 'id') {
            $id = $InputObject.id
        } else {
            $id = ""
        }

        if (Get-Member -InputObject $InputObject -Name 'subject') {
            $subject = $InputObject.subject
        } else {
            $subject = ""
        }

        if (Get-Member -InputObject $InputObject -Name 'timestamp') {
            $date = Get-Date -Date $InputObject.timestamp
        } else {
            $date = Get-Date
        }

        if (Get-Member -InputObject $InputObject -Name 'to') {
            $to = $InputObject.to
        } else {
            $to = ""
        }

        if (Get-Member -InputObject $InputObject -Name 'status') {
            $status = $InputObject.status
        } else {
            $status = "RECEIVED"
        }

        $email = [PSCustomObject]@{
            Date       = $date
            Status     = $status
            ID         = $id
            Client     = $client
            ClientPort = $clientPort
            From       = $from
            To         = $to
            Subject    = $subject
        }

        $email.PSObject.TypeNames.Insert(0, 'ArquebuseEmail')

        return $email
    }
}